### PR TITLE
Large deletion + changes to repair functions

### PR DESCRIPTION
**Pull request by Ellmen**
Big changes:
- addition of make_new_genome function
- target's repair now calls genome's repair_target directly
- large deletion functionality  
- genome simulation
- camv initialization  

Still TODO:
- ~add support for other 3 kinds of targets (backwards, other strand or both)~

Notes:
- this is still untested so it likely contains major bugs
- please take a close look at make_new_genome and see if it makes sense
- only need sense , not direction

### CODE REVIEW

**Comment by mattsmart** [comment]
I'll check this over for bugs sometime td

**Comment by mattsmart** [comment]
Everything looks OK so far, but I haven't gone through the big method `make_new_genome`

File: models/targeting/genome_classes.py [hunk 1]
```diff
@@ -212,3 +213,77 @@ def initialize_target_cut_probabilities(self, dt):
...
+            else:
+                # find new pam, set new location, set new sequence
+                new_start = self.find_pam(target.current_start)
+                self.domains[domain_label].targets[target.label].current_start = new_start
+                self.domains[domain_label].targets[target.label].sequence = self.current_genome[new_start:new_start+20]
+
+    def large_deletion(self, target1, target2):
+        """Delete section between two open targets
+        """
+        # make sure cut_positions of targets are up to date
```
Comment (mattsmart) [inline models/targeting/genome_classes.py:313]: maybe assert that they are cut (i.e. not repaired), first

**Comment by mattsmart** [comment]
aside from `make_new_genome`, what other parts may be sketchy?

File: models/targeting/genome_classes.py [hunk 2]
```diff
@@ -212,3 +213,77 @@ def initialize_target_cut_probabilities(self, dt):
...
+        target1.set_cut_position()
+        target2.set_cut_position()
+        if target1.cut_position > target2.cut_position:
+            location = target2.cut_position
+            del_size = target1.cut_position - target2.cut_position
+        else:
+            location = target1.cut_position
+            del_size = target2.cut_position - target1.cut_position
```
Comment (mattsmart) [inline models/targeting/genome_classes.py]: these 6 lines could be simplified to:
`location = min(target1.cut_position, target2.cut_position)`
`del_size = abs(target1.cut_position - target2.cut_position)`

Comment (Ellmen) [inline models/targeting/genome_classes.py]: good idea

File: models/targeting/genome_classes.py [hunk 3]
```diff
@@ -212,3 +213,74 @@ def initialize_target_cut_probabilities(self, dt):
...
+            # if domain starts past indel location
+            if domain.domain_start > indel_location:
+                # re-index it
+                self.domains[key_domain].domain_start += indel_size
+                # if this pulls start past indel location
+                if self.domains[key_domain].domain_start < indel_location:
+                    # set it to indel location
+                    self.domains[key_domain].domain_start = indel_location
+            # if domain ends past indel location
+            if domain.domain_end > indel_location:
```
Comment (mattsmart) [inline models/targeting/genome_classes.py:268]: should this be `elif`?

actually nvm I think

File: models/targeting/genome_classes.py [hunk 4]
```diff
@@ -212,3 +213,74 @@ def initialize_target_cut_probabilities(self, dt):
...
+        """Re-index all domains and targets after single indel
+        and set current_genome to new_genome
+
+        This is a bit headache inducing
+        """
+        deleted_domains = []
+        broken_targets = []
+        target_dict = self.get_targets_from_genome()
+        for key_domain in target_dict.keys():
+            domain = self.domains[key_domain]
```
Comment (mattsmart) [inline models/targeting/genome_classes.py:258]: Are you trying to use this `domain` as a copy to store some history? If so, we need to guarantee it's not acting as a pointer (I'm not sure), there are copy methods to correct it if it's a problem.

File: models/targeting/genome_classes.py [hunk 5]
```diff
@@ -212,3 +213,74 @@ def initialize_target_cut_probabilities(self, dt):
...
+                # if this pulls end past indel location
+                if self.domains[key_domain].domain_end < indel_location:
+                    # set it to indel location
+                    self.domains[key_domain].domain_end = indel_location
+                # if start and end are the same (both location)
+                if self.domains[key_domain].domain_start == self.domains[key_domain].domain_end:
+                    # it has been deleted
+                    deleted_domains.append(domain)
+                else:
+                    # otherwise check if it has targets to be re-indexed
```
Comment (mattsmart) [inline models/targeting/genome_classes.py:280]: looks OK up to here

File: models/targeting/genome_classes.py [hunk 6]
```diff
@@ -212,3 +213,74 @@ def initialize_target_cut_probabilities(self, dt):
...
+                            self.domains[key_domain].targets[key_target].current_start += indel_size
+                            # if the target has been damaged or deleted
+                            if self.domains[key_domain].targets[key_target].current_start < indel_location:
+                                # deal with that later
+                                broken_targets.append(self.domains[key_domain].targets[key_target])
+        # remove all deleted domains
+        for domain in deleted_domains:
+            self.remove_domain(domain)
+        # set new genome
+        self.current_genome = new_genome
```
Comment (mattsmart) [inline models/targeting/genome_classes.py:293]: should also update `self.length` and (maybe) `self.repaired` state

File: models/targeting/genome_classes.py [hunk 7]
```diff
@@ -212,3 +213,74 @@ def initialize_target_cut_probabilities(self, dt):
...
+        # delete or fix all broken targets
+        for target in broken_targets:
+            domain_label = target.domain.label
+            # if whole target is deleted
+            if target.current_start + 20 < location:
```
Comment (mattsmart) [inline models/targeting/genome_classes.py]: should this be `indel_location`, also 20 probably shouldn't be hardcoded but we'll leave it for now I think

File: models/targeting/genome_classes.py [hunk 8]
```diff
@@ -212,3 +213,74 @@ def initialize_target_cut_probabilities(self, dt):
...
+                self.domains[domain_label].remove_target(target)
+            # else it is just broken
+            else:
+                # find new pam, set new location, set new sequence
+                new_start = self.find_pam(target.current_start)
```
Comment (mattsmart) [inline models/targeting/genome_classes.py]: still a little iffy on that method

**Comment by mattsmart** [comment]
minor comments, let me know when you've pushed a new commit

looks pretty good, and clean considering how confusing it is lol

File: models/targeting/genome_classes.py (hunk 7 above)
Comment (Ellmen) [inline models/targeting/genome_classes.py]: oops, ya good catch

**Comment by mattsmart** [comment]
LGTM merging

**Pull request merged**